README.md
[![github latest release](https://badgen.net/github/release/nichelia/ekrhizoc?icon=github)](https://github.com/nichelia/ekrhizoc/releases/latest/)
[![pypi latest package](https://badgen.net/pypi/v/ekrhizoc?label=pypi%20pacakge)](https://pypi.org/project/ekrhizoc/)
[![docker latest image](https://img.shields.io/docker/v/nichelia/ekrhizoc?label=image&logo=docker&logoColor=white)](https://hub.docker.com/repository/docker/nichelia/ekrhizoc)
[![project license](https://badgen.net/github/license/nichelia/ekrhizoc?color=purple)](https://github.com/nichelia/ekrhizoc/blob/master/LICENSE)

![e6c CI](https://github.com/nichelia/ekrhizoc/workflows/e6c%20CI/badge.svg)
![e6c CD](https://github.com/nichelia/ekrhizoc/workflows/e6c%20CD/badge.svg)
[![security scan](https://badgen.net/dependabot/nichelia/ekrhizoc/?label=security%20scan)](https://github.com/nichelia/ekrhizoc/labels/security%20patch)
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen)](https://github.com/pre-commit/pre-commit)


[![code coverage](https://badgen.net/codecov/c/github/nichelia/ekrhizoc?label=code%20coverage)](https://codecov.io/gh/nichelia/ekrhizoc)
[![code alerts](https://badgen.net/lgtm/alerts/g/nichelia/ekrhizoc?label=code%20alerts)](https://lgtm.com/projects/g/nichelia/ekrhizoc/alerts/)
[![code quality](https://badgen.net/lgtm/grade/g/nichelia/ekrhizoc?label=code%20quality)](https://lgtm.com/projects/g/nichelia/ekrhizoc/context:python)
[![code style](https://badgen.net/badge/code%20style/black/color=black)](https://github.com/ambv/black)

# ekrhizoc
ekrhizoc (E6c): A web crawler

## Contents
1. [Definition](#definition)
2. [Use Case](#use-case)
3. [Configuration](#configuration)
4. [Development](#development)
5. [Testing](#testing)
6. [Versioning](#versioning)
7. [Deployment](#deployment)
8. [Production](#production)

## Definition

εκρίζωση (Greek)
ekrízosi / uprooting, eradication

Also known as __E6c__.

## Use Case

Implementation of a simple python web crawler.  
Input: URL (seed).  
Output: simple textual sitemap (to show links between pages).

### Requirements

* The crawler is limited to *__one__* subdomain (exclude external links).
* No use of web crawling libraries/frameworks (e.g. scrapy).
* (Optional) Use of HTML handling Libraries/Frameworks.
* Production-ready code.

### Assumptions

* The input URL (seed) is limited to __only__ one at every run.
* The targeted URL(s) are static pages (no backend javascript parsing required).
* Links to be extracted from HTML anchor `<a>` elements.
* Valid links include
    - Valid URL
        + Non empty
        + Matches a valid url pattern
        + Does not exceed the `E6C_MAX_URL_LENGTH` length in characters
        + Possible to convert a relative urls to a full url
    - Link is not visited before
    - Link is not part of an ignored file type
    - Link has the same domain as the seed url
    - Link is not restricted by the robots.txt file

### Design

This project implements a Basic Universal Crawler based on breadth first search graph traversal.

## Configuration

Behaviour of the application can be configured via Environment Variables.

| Environment Variable | Description | Type | Default Value |
| -------------- | -------------- | -------------- | -------------- |
| `E6C_LOG_LEVEL` | Level of logging - overrides verbose/quiet flag | string | - |
| `E6C_LOG_DIR` | Directory to save logs | string | - |
| `E6C_BIN_DIR` | Directory to save any output (bin) | string | bin |
| `E6C_IGNORE_FILETYPES` | File types of websites to ignore (e.g. ".filetype1,.filetype2") | string | ".png,.pdf,.txt,.doc,.jpg,.gif" |
| `E6C_URL_REQUEST_TIMER` | Time (in seconds) to wait per request (not to populate server with multiple requests) | float | 0.1 |
| `E6C_MAX_URLS` | The maximum number of urls to fetch/crawl | integer | 10000 |
| `E6C_MAX_URL_LENGTH` | The maximum length (character count) of a url to fetch/crawl | integer | 300 |

## Development

### Configure for local development

* Clone [repo](https://github.com/nichelia/ekrhizoc) on your local machine
* Install [`conda`](https://www.anaconda.com) or [`miniconda`](https://docs.conda.io/en/latest/miniconda.html)
* Create your local project environment (based on [`conda`](https://www.anaconda.com), [`poetry`](https://python-poetry.org), [`pre-commit`](https://pre-commit.com)):  
`$ make env`
* (Optional) Update existing local project environment:  
`$ make env-update`

### Run locally

On a terminal, run the following (execute on project's root directory):

* Activate project environment:  
`$ . ./scripts/helpers/environment.sh`
* Run the CLI using `poetry`:  
`$ poetry run ekrhizoc`

### Contribute

[ Not Available ]

## Testing
(part of CI/CD)

[ Work in progress... ]

To run the tests, open a terminal and run the following (execute on project's root directory):

* Activate project environment:  
`$ . ./scripts/helpers/environment.sh`
* To run pytest:  
`$ make test`
* To check test coverage:  
`$ make test-coverage`

## Versioning

Increment the version number:  
`$ poetry version {bump rule}`  
where valid bump rules are:

1. patch
2. minor
3. major
4. prepatch
5. preminor
6. premajor
7. prerelease

### Changelog

Use `CHANGELOG.md` to track the evolution of this package.  
The `[UNRELEASED]` tag at the top of the file should always be there to log the work until a release occurs.  

Work should be logged under one of the following subtitles:
* Added
* Changed
* Fixed
* Removed

On a release, a version of the following format should be added to all the current unreleased changes in the file.  
`## [major.minor.patch] - YYYY-MM-DD`

## Deployment

### Pip package

On a terminal, run the following (execute on project's root directory):

* Activate project environment:  
`$ . ./scripts/helpers/environment.sh`
* To build pip package:  
`$ make build-package`
* To publish pip package (requires credentials to PyPi):  
`$ make publish-package`

### Docker image

On a terminal, run the following (execute on project's root directory):

* Activate project environment:  
`$ . ./scripts/helpers/environment.sh`
* To build docker image:  
`$ make build-docker`

## Production

For production, a Docker image is used.
This image is published publicly on [docker hub](https://hub.docker.com/repository/docker/nichelia/ekrhizoc).

* First pull image from docker hub:  
`$ docker pull nichelia/ekrhizoc:[version]`
* First pull image from docker hub:  
`$ docker run --rm -it -v ~/ekrhizoc_bin:/usr/src/bin nichelia/ekrhizoc:[version]`  
This command mounts the application's bin (outcome) to user's root directory under ekrhizoc_bin folder.

where version is the published application version (e.g. 0.1.0)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
- Main Docker filename

## [0.1.1] - 2020-02-25
### Added
- Reconfigure CD with Github Actions (includes: package, publish, release)
- Reconfigure CI with Github Actions (includes: lint, test, security scan)
- Support for pre-commit
- GitHub issue templates for Bug report and Feature request
- Badges for the project in README
- Setup CI with Travis
- Test coverage support with Codecov
- Pre-commit hooks
- Pylint / shellcheck code fixes

## [0.1.0] - 2020-04-12
### Added
- More debug logging for (in)valid URL
- Tests for canonical urls
- Tests for (in)valid urls
- Tests for extracting url domain
- Tests for detecting same url domain
- Tests for crawlers detection
- Tests for commands detection
- Tests for crawler command

### Changed
- Regex for valid URL
- Poetry project details
- Updated README

### Fixed
- Logger bug, where log level would not behave as expected if command verbosity level and environment variable are set

## [0.0.5] - 2020-02-24
### Added
- Project skeleton: setup conda for project environment, setup poetry for project dependencies, setup githooks
- Makefile, Bash scripts to handle project setup
- License, README, CHANGELOG
- Docker setup
- Settings, Logs configuration
- Code comments
- CLI setup, available commands: crawl with verbose
- Crawler: Universal breadth first search graph traversal
- Crawler features: Asynchronous calls, page url scraper, same domain url filter, visited urls filter, file type filtering, url length limits, handle relative urls, respect robots.txt file
- Output as graph

[Unreleased]: https://github.com/nichelia/ekrhizoc/compare/0.1.1...HEAD
[0.1.1]: https://github.com/nichelia/ekrhizoc/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/nichelia/ekrhizoc/compare/0.0.5...0.1.0
[0.0.5]: https://github.com/nichelia/ekrhizoc/releases/tag/0.0.5


ekrhizoc/settings.py
from os import getenv
from typing import Any


class EkrhizocSetting:
    def __init__(
        self,
        field: str,
        datatype: type,
        default_value: Any,
        env_var_alias: str = None,
        description: str = None,
    ):
        self.field = field
        self.datatype = datatype
        self.default_value = default_value
        self.env_var_alias = env_var_alias
        self.description = description
        self._name = "_".join(("E6C", self.field.upper()))

    def __invert__(self) -> Any:
        """ Helper function.

        For convinience, access the setting value
        by using the invert ~.
        e.g. ~EkrhizocSetting == EkrhizocSetting.value

        Returns:
            The value of the value attribute.
        """
        return self.value

    @property
    def name(self) -> str:
        """Getter for name attribute.

        Returns:
            The value of the name attribute.
        """
        return self._name

    @property
    def env_var(self) -> Any:
        """Getter for env_var attribute.

        Check for a value in the environment variable,
        or the alias environment variable.


        Returns:
            The value of the env_var attribute (default None).
        """
        return getenv(
            self._name,
            getenv(self.env_var_alias, None) if self.env_var_alias else None,
        )

    @property
    def value(self) -> Any:
        """Getter for value attribute.

        If a value exists in an environment variable,
        return it otherwise return the default value.

        Returns:
            The value of the value attribute.
        """
        if self.env_var:
            return self.datatype(self.env_var)
        return self.default_value


LOG_LEVEL = EkrhizocSetting(
    field="log_level",
    datatype=str,
    default_value="",
    description="Level of logging - overrides verbose flag",
)

LOG_DIR = EkrhizocSetting(
    field="log_dir",
    datatype=str,
    default_value="",
    description="Directory to save logs",
)

BIN_DIR = EkrhizocSetting(
    field="bin_dir",
    datatype=str,
    default_value="bin",
    description="Directory to save any output (bin)",
)

IGNORE_FILETYPES = EkrhizocSetting(
    field="ignore_filetypes",
    datatype=str,
    default_value=".png,.pdf,.txt,.doc,.jpg,.gif",
    description='File types of websites to ignore (e.g. ".filetype1,.filetype2")',
)

URL_REQUEST_TIMER = EkrhizocSetting(
    field="url_request_timer",
    datatype=float,
    default_value=0.1,
    description="Time (in seconds) to wait per request (not to populate server with multiple requests)",
)

MAX_URLS = EkrhizocSetting(
    field="max_urls",
    datatype=int,
    default_value=10000,
    description="The maximum number of urls to fetch/crawl",
)

MAX_URL_LENGTH = EkrhizocSetting(
    field="max_url_length",
    datatype=int,
    default_value=300,
    description="The maximum length (character count) of a url to fetch/crawl",
)
